WorkflowSwiftUI/Sources/Store.swift

```diff
@@ -0,0 +1,453 @@
+///
+/// Because arbitrary properties on the model cannot be tracked for observation, any other types of
+/// properties will not be accessible.
+///
+/// For state properties that are writable, an automatic `Binding` can be derived by annotating the
+/// store with `@Bindable`. These bindings will use the workflow's state mutation sink.
+///
+/// All properties can be turned into bindings by appending `sending(store:action:)` or
+/// `sending(closure:)` to specify the "write" action. For properties that are already writable,
+/// this will refine the binding to send a custom action instead of the built-in state mutation
+/// sink.
+///
+@dynamicMemberLookup
+public final class Store<Model: ObservableModel>: Perceptible {
+    public typealias State = Model.State
+
+    private var model: Model
+    private let _$observationRegistrar = PerceptionRegistrar()
+
+    private var childStores: [AnyHashable: ChildStore] = [:]
+    private var childModelAccesses: [AnyHashable: ChildModelAccess] = [:]
+    private var invalidated = false
+
+    static func make(model: Model) -> (Store, (Model) -> Void) {
+        let store = Store(model)
+        return (store, store.setModel)
+    }
+
+    fileprivate init(_ model: Model) {
+        self.model = model
+    }
+
+    var state: State {
+        _$observationRegistrar.access(self, keyPath: \.state)
+        return model.accessor.state
+    }
+
+    private func send<Value>(keyPath: WritableKeyPath<State, Value>, value: Value) {
+        guard !invalidated else {
+            return
+        }
+        model.accessor.sendValue { state in
+            state[keyPath: keyPath] = value
+        }
+    }
+
+    fileprivate func setModel(_ newModel: Model) {
+        // Make a list of any child store accesses that are mutated as a result of this set. We'll
+        // use this list to wrap the update with appriate willSet/didSet calls.
+        let changedChildAccess = childModelAccesses.values.filter { $0.changed(model, newModel) }
+
+        /// Update the model, wrapped in willSet and didSet observations for mutations to child
+        /// store wrappers.
+        func updateModel() {
+            for access in changedChildAccess {
+                access.willSet(self)
+            }
+
+            model = newModel
+
+            for access in changedChildAccess {
+                access.didSet(self)
+            }
+        }
+
+        // Update the model, registering a mutation if the state has changed
+
+        if !_$isIdentityEqual(model.accessor.state, newModel.accessor.state) {
+            _$observationRegistrar.withMutation(of: self, keyPath: \.state) {
+                updateModel()
+            }
+        } else {
+            updateModel()
+        }
+
+        // Update and invalidate child stores
+
+        for (keyPath, childStore) in childStores {
+            if childStore.isInvalid(newModel) {
+                childStore.invalidate()
+                childStores[keyPath] = nil
```

Comment:
Does this have any performance implications? Eg, does this force a redraw of any child screen / nested references?

Reply:
Nope, releasing a cached child store shouldn't affect views at all. Cached child stores are invalidated if they're no longer accessible, e.g. if a collection changed and the associated element is no longer present. This prevents the cache from growing unbounded.